api/test/wfl/unit/source_test.clj

```diff
@@ -89,3 +91,37 @@
       (is (nil? (#'source/filter-and-combine-tdr-source-details
                  [failed]))
           "failed rows should not be kept"))))
+
+(deftest test-check-tdr-job-and-notify-on-failure
+  (let [workload    {}
+        job-id      "tdr-job-id"
+        job-result  {:id "tdr-snapshot-id"}]
+    (letfn [(mock-tdr-job-failed-slack-msg
+              [{:keys [job_status] :as _metadata} _caught-job-result]
+              (is (not (#{"running" "succeeded"} job_status))
+                  "Should not notify for running or succeeded jobs"))]
+      (with-redefs
+       [datarepo/job-result             (constantly job-result)
+        source/tdr-job-failed-slack-msg mock-tdr-job-failed-slack-msg
+        slack/notify-watchers           (constantly nil)]
+        (let [metadata {:job_status "running"}]
+          (with-redefs [datarepo/job-metadata (constantly metadata)]
```

Comment:
Since you brought my attention here, I do find `with-redefs` easier to work with and read than `with-redefs-fn`, especially in cases where wanting to nest redefinitions.

Reply:
Glad to hear it.